File: hospital/SergioDeIscar-hospital-java/SergioDeIscar-hospital-java/src/main/java/org/example/repositories/HospitalRepoConjunto.java
package org.example.repositories;

import org.example.enums.TipoPaciente;
import org.example.interfaces.HospitalExtension;
import org.example.models.Paciente;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

public class HospitalRepoConjunto implements HospitalExtension {
    private final int maxSize;
    public HospitalRepoConjunto() {
        this.maxSize = 50;
    }
    public HospitalRepoConjunto(int maxSize) {
        this.maxSize = maxSize;
    }

    Set<Paciente> pacientes = new LinkedHashSet<>();

    @Override
    public Paciente save(Paciente entity) {
        if (estaCompleto() || entity == null) return null;
        entity.setFechaIngreso(LocalDate.now());
        pacientes.add(entity);
        return entity;
    }

    @Override
    public Paciente findById(String s) {
        return pacientes.stream().filter(p -> p.getDni().equals(s)).findFirst().orElse(null);
    }

    @Override
    public Paciente deleteById(String s) {
        Paciente paciente = findById(s);
        if (paciente == null) return null;
        paciente.setFechaAlta(LocalDate.now());
        pacientes.remove(paciente);
        return paciente;
    }

    @Override
    public List<Paciente> findAll() {
        return pacientes.stream().toList();
    }

    @Override
    public void saveAll(List<Paciente> entities) {
        if (estaCompleto()) return;
        entities.forEach(p -> save(p));
    }

    @Override
    public Boolean estaCompleto() {
        return pacientes.size() >= maxSize;
    }

    @Override
    public int numPacientes() {
        return pacientes.size();
    }

    @Override
    public List<Paciente> pacientesOrdeFechaIngreso() {
        return pacientes.stream().sorted((p1, p2) -> p1.getFechaIngreso().compareTo(p2.getFechaIngreso())).toList();
    }

    @Override
    public List<Paciente> pacientesOrdeNombre() {
        return pacientes.stream().sorted((p1, p2) -> p1.getNombre().compareTo(p2.getNombre())).toList();
    }

    @Override
    public List<Paciente> pacientesOrdeDNI() {
        return pacientes.stream().sorted((p1, p2) -> p1.getDni().compareTo(p2.getDni())).toList();
    }

    @Override
    public List<Paciente> pacientesPorTipo(TipoPaciente tipo) {
        return pacientes.stream().filter(p -> p.getTipo().equals(tipo)).toList();
    }

    @Override
    public int numPacientePorTipo(TipoPaciente tipo) {
        return pacientesPorTipo(tipo).size();
    }

    @Override
    public int capacidadMaxima() {
        return maxSize;
    }
}

File: hospital/SergioDeIscar-hospital-kotlin/SergioDeIscar-hospital-kotlin/src/main/kotlin/repositories/HospitalRepoConjunto.kt
package repositories

import enums.TipoPaciente
import interfaces.HospitalExtension
import models.Paciente
import java.time.LocalDate

class HospitalRepoConjunto(private val maxSize: Int = 50): HospitalExtension {
    // LinkedHashSet, aunque me lo pidan ordenado hacerlo con un TreeSet ralentiza el proceso
    // y lo puedo devolver ordenado cuando me lo pidan
    private val pacientes = mutableSetOf<Paciente>()

    override fun estaCompleto(): Boolean {
        return pacientes.size >= maxSize
    }

    override fun numPacientes(): Int {
        return pacientes.size
    }

    override fun pacientesOrdeFechaIngreso(): List<Paciente> {
        return pacientes.sortedBy { it.fechaIgreso }
    }

    override fun pacientesOrdeNombre(): List<Paciente> {
        return pacientes.sortedBy { it.nombre }
    }

    override fun pacientesOrdeDNI(): List<Paciente> {
        return pacientes.sortedBy { it.dni }
    }

    override fun pacientesPorTipo(tipo: TipoPaciente): List<Paciente> {
        //Dos formas de hacerlo
        return pacientes.filter { it.tipo == tipo }
        /*val list = mutableListOf<Paciente>()
        for (i in pacientes){
            if (i.tipo == tipo) list.add(i)
        }
        return list.toList()*/
    }

    override fun numPacientePorTipo(tipo: TipoPaciente): Int {
        return pacientesPorTipo(tipo).count()
    }

    /**
     * Si el paciente ya existe lo actualiza, si no lo añade
     */
    override fun save(paciente: Paciente): Paciente? {
        if (estaCompleto()) return null
        paciente.setFechaIngreso(LocalDate.now())
        pacientes.add(paciente)
        return paciente
    }

    override fun find(dni: String): Paciente? {
        // Programación funcional, mejor modo de hacerlo
        return pacientes.find { it.dni == dni }

        // For each, mas lento
        /*for (i in pacientes){
            if (i.dni == dni) return i
        }
        return null*/
    }

    override fun delete(dni: String): Paciente? {
        val paciente = find(dni) ?: return null
        paciente.setFechaAlta(LocalDate.now())
        pacientes.remove(paciente)
        return paciente
    }

    override fun getAll(): List<Paciente> {
        return pacientes.toList()
    }

    override fun saveAll(pacientes: List<Paciente>) {
        if (estaCompleto()) return
        pacientes.forEach{ save(it) }
    }
}